activerecord/lib/active_record/associations/association.rb

```diff
@@ -409,11 +409,13 @@ def inversable?(record)
         end
 
         def matches_foreign_key?(record)
-          if foreign_key_for?(record)
-            record.read_attribute(reflection.foreign_key) == owner.id ||
-              (foreign_key_for?(owner) && owner.read_attribute(reflection.foreign_key) == record.id)
```

Comment:
I'm really not sure what this logic here was trying to achieve, but it for sure didn't seem right. Either it was missing test cases, or it was never utilized. Either way, all tests pass with the reworked logic I added.

Reply:
Is it possible to maintain something similar so that it avoids allocating arrays in the case of a single pkey?